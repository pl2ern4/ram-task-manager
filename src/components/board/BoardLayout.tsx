
import {useState} from 'react';
import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import Paper from '@mui/material/Paper';
import Card from '../card/Card';
import { styled } from '@mui/material/styles';
import FormDialog from '../dialog/Dialog';
import Task from '../task/Task';
import { TICKET_STATUS, TICKET_ORDERBY } from '../../constants';
import CreateBoard from '../create-task/CreateTask';
import { useGlobalContext } from '@/context/application';
import SelectInput from '../select-input/SelectInput';

const ColumnBackground = styled(Paper)`
  width: calc(100vw/4);
  padding: .5rem;
  min-height: calc(100vh - 10rem);
  background-color: #8080801f;
`;
const BoardWrapper = styled('div')`
  width: calc(100vw - 2rem);
  justify-content: center;
`;

export default function BoardLayout() {
  const [ticketOrder, setTicketOrder] = useState('modifiedDate');
  const {openCreateTicket, openUpdateTicketDialog, openUpdateTicket, setOpenCreateTicket, tickets, updateTicket} = useGlobalContext();
  const TaskColumn = TICKET_STATUS;

  const onDragEnd = (evt) => {
    evt.currentTarget.classList.remove("dragged");
  };

  const onDragStart = (evt) => {
    let element = evt.currentTarget;
    element.classList.add("dragged");
    evt.dataTransfer.setData("draggedItem", evt.currentTarget.id);
    evt.dataTransfer.effectAllowed = "move";
  };

  const onDragLeave = (evt)=>{
    evt.currentTarget.classList.remove("dragged");
  }
  const onDragEnter = (evt)=>{
    evt.preventDefault();
    let element = evt.currentTarget;
    element.classList.add("dragged-over");
    evt.dataTransfer.dropEffect = "move";
  }

  const onDragOver = (evt)=>{
    evt.preventDefault();
    evt.dataTransfer.dropEffect = "move";
  }

  const onDrop = (evt)=>{
    evt.preventDefault();
    evt.currentTarget.classList.remove("dragged-over");
    const draggedItemId = evt.dataTransfer.getData("draggedItem");
    const index = tickets.findIndex(o=>o.id==draggedItemId)
    if(index>-1 && evt.target.id){
      updateTicket({id: draggedItemId, node:"status", data: evt.target.id})
    }
  }
  const handleSortChange = (_, value)=>{
    setTicketOrder(value)
  }

  const modiefieldTicketList = ({list, orderBy, value})=>{
    if(orderBy==='bookmarked'){
      return list.reduce((acc, item)=>item.isBookMarked && [item,...acc] || [...acc, item],[]).filter(o=>o.status===value);
    }
    if(orderBy==='createdDate'){
      return list.sort((a,b)=>a.createdDate>b.createdDate?-1:1).filter(o=>o.status===value);
    }
    if(orderBy==='title'){
      return list.sort((a,b)=>a.title>b.title?1:-1).filter(o=>o.status===value);
    }
    return list.sort((a,b)=>a.modifiedDate>b.modifiedDate?-1:1).filter(o=>o.status===value);
  }
  
  return (
    <BoardWrapper>
      <SelectInput onChange={handleSortChange} defaultOption={ticketOrder} options={Object.keys(TICKET_ORDERBY)}  label="Order By"/>
      <Grid sx={{ flexGrow: 1, justifyContent:'center' }} container justifyContent="center" spacing={2}>
        <Grid item xs={12}>
          <Grid container justifyContent="center" spacing={2} sx={{justifyContent:'center' }}>
            {TaskColumn.map((value) => (
              <Grid key={value} item>
                <ColumnBackground 
                  className="drop-zone order small-box" 
                  id={value} 
                  onDragLeave={onDragLeave} 
                  onDragEnter={onDragEnter}
                  onDragEnd={onDragEnd}
                  onDragOver={onDragOver}
                  onDrop={onDrop}>
                  <Typography variant="h5" color="text.primary" gutterBottom>
                    {value}
                  </Typography>
                    {
                      modiefieldTicketList({list:tickets, orderBy:ticketOrder, value}).map(item=> <Card onDragStart={onDragStart} onDragEnd={onDragEnd} key={item.id} isBookMarked={item.isBookMarked} id={item.id} title={item.title} storyPoint={item.storyPoint} assigneeName={item.asignee} taskType={item.taskType} />)
                    }
                </ColumnBackground>
              </Grid>
            ))}
          </Grid>
        </Grid>
      </Grid>
      {openCreateTicket && <FormDialog openDialog={openCreateTicket} onClose={shouldOpenDialog=>setOpenCreateTicket(shouldOpenDialog)}><CreateBoard/></FormDialog>}
      {openUpdateTicketDialog.shouldOpenDialog && <FormDialog openDialog={openUpdateTicketDialog.shouldOpenDialog} onClose={(shouldOpenDialog)=>openUpdateTicket(undefined, shouldOpenDialog)}><Task/></FormDialog>}
    </BoardWrapper>

  );
}