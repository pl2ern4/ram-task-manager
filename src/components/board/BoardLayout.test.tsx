import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import BoardLayout from './BoardLayout';

describe('BoardLayout', () => {
  it('should Match Snapshot', () => {
    const { container } = render(<BoardLayout />)
    expect(container).toMatchSnapshot()
  })
})