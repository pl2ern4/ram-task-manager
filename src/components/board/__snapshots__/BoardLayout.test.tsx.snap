// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardLayout should Match Snapshot 1`] = `
<div>
  <div
    class="css-33ztoe"
  >
    <div
      class="css-1dw64bn"
    >
      <label
        class="css-14pxzl7"
      >
        Order By
      </label>
      <button
        aria-controls=":r0:"
        aria-expanded="false"
        class="base-Select-root css-12xgkxt"
        role="combobox"
        type="button"
      >
        <span
          class="notranslate"
        >
          â€‹
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="UnfoldMoreRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m12 5.83 2.46 2.46c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L12.7 3.7a.9959.9959 0 0 0-1.41 0L8.12 6.88c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0zm0 12.34-2.46-2.46a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41l3.17 3.18c.39.39 1.02.39 1.41 0l3.17-3.17c.39-.39.39-1.02 0-1.41a.9959.9959 0 0 0-1.41 0z"
          />
        </svg>
      </button>
      <input
        aria-hidden="true"
        style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
        tabindex="-1"
        value=""
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-zrswig-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 drop-zone order small-box css-92034v-MuiPaper-root"
              id="Todo"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1rwkfxe-MuiTypography-root"
              >
                Todo
              </h5>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 drop-zone order small-box css-92034v-MuiPaper-root"
              id="inProgress"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1rwkfxe-MuiTypography-root"
              >
                inProgress
              </h5>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 drop-zone order small-box css-92034v-MuiPaper-root"
              id="Done"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1rwkfxe-MuiTypography-root"
              >
                Done
              </h5>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
